axon/threads_test.go

```diff
@@ -0,0 +1,353 @@
+					fieldS := synS.Field(fi)
+					fieldM := synM.Field(fi)
+					if fieldS.Kind() == reflect.Float32 {
+						assert.Equal(t, fieldS.Float(), fieldM.Float(),
+							"Synapse %d, field %s, single thread: %f, multi thread: %f",
+							si, synS.Type().Field(fi).Name, fieldS.Float(), fieldM.Float())
+					} else if fieldS.Kind() == reflect.Int32 {
+						assert.Equal(t, fieldS.Int(), fieldM.Int(),
+							"Synapse %d, field %s, single thread: %d, multi thread: %d",
+							si, synS.Type().Field(fi).Name, fieldS.Int(), fieldM.Int())
+					}
+				}
+			}
+		}
+	}
+
+	require.Equalf(t, 0.0, maxDiff.diff,
+		"Max difference (floats only): %f, field %s, neuron %d, single thread: %f, multi thread: %f",
+		maxDiff.diff, maxDiff.field, maxDiff.neuronIdx, maxDiff.valShould, maxDiff.valIs)
+}
+
+// I search for a while for a good way to have expected test failures in Go, but
+// couldn't find anything, so I wrote this additional function that returns a bool
+// that I can check. Please let me know if there is a better way to do this.
+func neuronsAreEqual(netS *Network, netM *Network) bool {
+	for li := range netS.Layers {
+		layerS := netS.Layers[li].(AxonLayer).AsAxon()
+		layerM := netM.Layers[li].(AxonLayer).AsAxon()
+		for ni := range layerS.Neurons {
+			nrnS := reflect.ValueOf(layerS.Neurons[ni])
+			nrnM := reflect.ValueOf(layerM.Neurons[ni])
+			for fi := 0; fi < nrnS.NumField(); fi++ {
+				fieldS := nrnS.Field(fi)
+				fieldM := nrnM.Field(fi)
+				if fieldS.Kind() == reflect.Float32 {
+					if fieldS.Float() != fieldM.Float() {
+						return false
+					}
+				} else if fieldS.Kind() == reflect.Int32 {
+					if fieldS.Int() != fieldM.Int() {
+						return false
+					}
+				}
+			}
+		}
+	}
+	return true
+}
+
+func generateRandomPatterns(nPats int) *etable.Table {
+	shape := []int{shape1D, shape1D}
+
+	rand.Seed(42)
+
+	pats := &etable.Table{}
+	pats.SetFromSchema(etable.Schema{
+		{Name: "Name", Type: etensor.STRING, CellShape: nil, DimNames: nil},
+		{Name: "Input", Type: etensor.FLOAT32, CellShape: shape, DimNames: []string{"Y", "X"}},
+		{Name: "Output", Type: etensor.FLOAT32, CellShape: shape, DimNames: []string{"Y", "X"}},
+	}, nPats)
+	numOn := shape[0] * shape[1] / 8
+	patgen.PermutedBinaryRows(pats.Cols[1], numOn, 1, 0)
+	patgen.PermutedBinaryRows(pats.Cols[2], numOn, 1, 0)
+	return pats
+}
+
+// buildIdenticalNetworks builds two identical nets, one single-threaded and one
+// multi-threaded. They are seeded with the same RNG, so they are identical.
+// Returns two networks: (single-threaded, multi-threaded)
+func buildIdenticalNetworks(t *testing.T, pats *etable.Table, tNeuron, tSendSpike, tSynCa int) (*Network, *Network) {
+	shape := []int{shape1D, shape1D}
+
+	// single-threaded network
+	rand.Seed(1337)
+	netS := buildNet(t, shape, 1, 1, 1)
+	// multi-threaded network
+	rand.Seed(1337)
+	netM := buildNet(t, shape, tNeuron, tSendSpike, tSynCa)
+
+	// The below code doesn't work, because we have no clean way of storing and restoring
```

Comment:
unless we have an issue tracking making this work, just delete the code, maybe replace it with a comment describing the general approach that won't work